MoreLinq/MaxBy.cs

```diff
@@ -81,29 +81,74 @@ static partial class MoreEnumerable
         // >
         // > - https://en.wikipedia.org/wiki/Maxima_and_minima
 
-        static TSource ExtremumBy<TSource, TKey>(IEnumerable<TSource> source,
+        static IEnumerable<TSource> ExtremaBy<TSource, TKey>(IEnumerable<TSource> source,
             Func<TSource, TKey> selector, Func<TKey, TKey, int> comparer)
         {
-            using (var sourceIterator = source.GetEnumerator())
+            return new LazyList<TSource>(new Lazy<List<TSource>>(() =>
             {
-                if (!sourceIterator.MoveNext())
-                    throw new InvalidOperationException("Sequence contains no elements");
-
-                var extremum = sourceIterator.Current;
-                var key = selector(extremum);
-                while (sourceIterator.MoveNext())
+                using (var e = source.GetEnumerator())
                 {
-                    var candidate = sourceIterator.Current;
-                    var candidateProjected = selector(candidate);
-                    if (comparer(candidateProjected, key) > 0)
+                    if (!e.MoveNext())
+                        return new List<TSource>();
+
+                    var extrema = new List<TSource> { e.Current };
+                    var extremaKey = selector(e.Current);
+
+                    while (e.MoveNext())
                     {
-                        extremum = candidate;
-                        key = candidateProjected;
+                        var item = e.Current;
+                        var key = selector(item);
+                        var comparison = comparer(key, extremaKey);
+                        if (comparison > 0)
+                        {
+                            extrema = new List<TSource> { item };
```

Comment:
Is there a reason for not using `extrema.Clear()` + `extrema.Add(item)` ? Using this could save some allocations and GC pressure, at the cost of potentially keeping more memory allocated than strictly necessary (ie, if there are many equal keys that appear before the true maximal key).

Reply:
> at the cost of potentially keeping more memory allocated than strictly necessary

I think you answered the question yourself.

The cost is going to be different for everyone depending on what's going to be their most common case. If you've collected the wrong list of extrema and going to restart it then `Clear` is going to spend time zero-ing out the internal array to release the elements. You then also risk, under memory pressure conditions, promoting the list and the internal array to the next generation. If you use `TrimExcess` then you still end up with an allocation and so you're only saving the trivial cost of a `List<>` allocation. The simplest thing to do therefore is to let the system do the balancing act and start afresh with a new list each time it needs to be restarted. The clever thing to do would be to build an adaptive system, but I don't think we want to get into that level of complexity (YAGNI?). Even `OrderBy` is hard-wired to one sorting algorithm.